src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -5701,6 +5701,60 @@ CancellationCode StorageReplicatedMergeTree::killMutation(const String & mutatio
     return CancellationCode::CancelSent;
 }
 
+void StorageReplicatedMergeTree::removePartsFromFilesystem(const DataPartsVector & parts)
+{
+    const auto settings = getSettings();
+    if (settings->max_part_removal_threads > 1 && parts.size() > settings->concurrent_part_removal_threshold)
+    {
+        /// Parallel parts removal.
+
+        size_t num_threads = std::min(size_t(settings->max_part_removal_threads), parts.size());
+        ThreadPool pool(num_threads);
+
+        /// NOTE: Under heavy system load you may get "Cannot schedule a task" from ThreadPool.
+        for (const DataPartPtr & part : parts)
+        {
+            pool.scheduleOrThrowOnError([&, thread_group = CurrentThread::getGroup()]
+            {
+                SCOPE_EXIT_SAFE(
+                    if (thread_group)
+                        CurrentThread::detachQueryIfNotDetached();
+                );
+                if (thread_group)
+                    CurrentThread::attachTo(thread_group);
+
+                LOG_DEBUG(log, "Removing part from filesystem {}", part->name);
+                try
+                {
+                    bool keep_s3 = !this->unlockSharedData(*part);
```

Comment:
Why variable called keep_s3, does merge need need to care where part is actually stored ?

Reply:
I'm not familiar with this part, it was just moved into separate helper (looks pretty safe)

Reply:
I see, better to keep previous code that worked before.